hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +85,964 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+                                                     access_location::host,
+                                                     access_mode::read);
+            ArrayHandle<unsigned int> h_molecule_list(m_rigid_bodies->getMoleculeList(),
+                                                      access_location::host,
+                                                      access_mode::read);
+            Index2D molecule_indexer = m_rigid_bodies->getMoleculeIndexer();
+            unsigned int nmol = molecule_indexer.getH();
+            // access body definitions
+            ArrayHandle<Scalar3> h_body_pos(m_rigid_bodies->getBodyOffsets(),
+                                            access_location::host,
+                                            access_mode::read);
+            Index2D h_body_idx = m_rigid_bodies->getBodyIndexer();
+            // accesss particle data
+            ArrayHandle<Scalar4> h_postype(m_pdata->getPositions(),
+                                           access_location::host,
+                                           access_mode::read);
+            ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                            access_location::host,
+                                            access_mode::read);
+            ArrayHandle<unsigned int> h_body(m_pdata->getBodies(),
+                                             access_location::host,
+                                             access_mode::read);
+            ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(),
+                                             access_location::host,
+                                             access_mode::read);
+            // loop over all molecules
+            unsigned int n_particles_local = m_pdata->getN() + m_pdata->getNGhosts();
+            for (unsigned int ibody = 0;
+                 ibody < nmol && !(invalid_center_of_mass || rigid_types.empty());
+                 ibody++)
+                {
+                // get central particle tag from first particle in molecule
+                assert(h_molecule_len.data[ibody] > 0);
+                unsigned int first_idx = h_molecule_list.data[molecule_indexer(0, ibody)];
+                if (first_idx > n_particles_local)
+                    continue;
+                unsigned int central_tag = h_body.data[first_idx];
+                unsigned int central_idx = h_rtag.data[central_tag];
+                if (central_idx >= n_particles_local)
+                    continue;
+
+                // only do molecules whose type hasn't been checked before
+                unsigned int type = __scalar_as_int(h_postype.data[central_idx].w);
+                if (rigid_types.find(type) == rigid_types.end())
+                    {
+                    continue;
+                    }
+                else
+                    {
+                    rigid_types.erase(rigid_types.find(type));
+                    }
```

Comment:
I guess the reason you're looping over particles, rather than body definitions, is that the mass is only available in the particle data (and not the body definition)? In that case, I think you probably should not check only the first instance of a body of a given type, but rather any whose type is in the set.
```suggestion
                // only do molecules participating in the collision
                unsigned int type = __scalar_as_int(h_postype.data[central_idx].w);
                if (rigid_types.find(type) == rigid_types.end())
                    {
                    continue;
                    }
```